solr/core/src/java/org/apache/solr/handler/ReplicationHandler.java

```diff
@@ -1241,33 +1241,33 @@ public void inform(SolrCore core) {
       numberBackupsToKeep = 0;
     }
     @SuppressWarnings({"rawtypes"})
-    NamedList slave = (NamedList) initArgs.get("slave");
-    boolean enableSlave = isEnabled( slave );
-    if (enableSlave) {
-      currentIndexFetcher = pollingIndexFetcher = new IndexFetcher(slave, this, core);
-      setupPolling((String) slave.get(POLL_INTERVAL));
-      isSlave = true;
+    NamedList secondary = (NamedList) initArgs.get("secondary");
+    boolean enableSecondary = isEnabled( secondary );
+    if (enableSecondary) {
+      currentIndexFetcher = pollingIndexFetcher = new IndexFetcher(secondary, this, core);
+      setupPolling((String) secondary.get(POLL_INTERVAL));
+      isSecondary = true;
     }
     @SuppressWarnings({"rawtypes"})
-    NamedList master = (NamedList) initArgs.get("master");
-    boolean enableMaster = isEnabled( master );
+    NamedList primary = (NamedList) initArgs.get("primary");
```

Comment:
Isn't this a communication between different Solr nodes?
If the current node got updated to new vocabulary primary/secondary but the remote node did not and did send data using master/slave, we might run into issues.

It might be desirable (if the above is true) to support both primary and master as return values (and secondary and slave) until a subsequent version of Solr not expected to work with nodes running code using master/slave.

Reply:
Also same issue in the other direction: if an updated node sends "primary" to a non updated node expecting "master", the non updated node will fail. Maybe we need to send the same data under both "primary" and "master" and the same data under both "secondary" and "slave", and accept either one on the receiving end.

I'm not familiar with how this code exactly works so might be off here, but this rename is not internal so must use caution (and test with renamed nodes talking to non renamed nodes, on both roles).

Reply:
I'm going to look through it. Hopefully I can grok it. Not sure, though. If I can find a more backcompat approach, I will drop it in the PR. Open to suggestions from others.

Reply:
`initArgs.containsKey("primary") ? initArgs.get("primary") : initArgs.get("master")`?

Reply:
As @chatman commented in the general discussion, maybe these changes happening in a new major version branch does no require them to be backward/forward compatible?

If 8x and 9.0 servers are expected to talk and understand each other, the suggestion looks good on the receiving end. We need something on the sending end as well (send both values, because "primary" being sent to a 8x server expecting only "master" will not be understood).

Reply:
I changed it to this: `NamedList primary = (NamedList)(initArgs.get("primary") != null ? initArgs.get("primary") : initArgs.get("master"));`

Reply:
We still promote rolling upgrade as the way to do upgrades in https://lucene.apache.org/solr/guide/8_6/upgrading-a-solr-cluster.html with no disclaimer about major versions. In the past we have strived to make a new major version be able to talk to previous major version. Could also let 8.7 support both and have the back-compat and require users to upgrade to 8.7 or later 8.x before doing rolling upgrade to 9.0. Or document that 8.x to 9.0 will not support rolling upgrade...

Reply:
Rolling upgrade across major versions isn't supported, and I don't see any reason to bend over backwards to support that. We can just document the changes and move on. This PR should be merged as soon as possible because it is a very large PR and can fall out of date very quickly.

Reply:
> Rolling upgrade across major versions isn't supported
I hope that is clearly documented somewhere? The only refguide page I could find on upgrading is the rolling one :)

Btw - since when was the size of a PR a good reason for rushing it into the codebase?

Reply:
>  I hope that is clearly documented somewhere? The only refguide page I could find on upgrading is the rolling one :)

If not documented, it should be. It has never been supported explicitly.

> Btw - since when was the size of a PR a good reason for rushing it into the codebase?

Hahah ;-) Ever since you displayed intent to sabotage/stall this effort (by requesting Marcus to go back to the solr-user list and discussing it there) :-) Anyway, jokes aside, this PR is quite close to being ready, except for the backcompat issue (which I don't think matters at all).

Reply:
> except for the backcompat issue (which I don't think matters at all).

What's the real issue? Let's accept as input "master" or "primary" in one case and "slave" or "secondary" in the other, as well as send both values over the wire as output.
This will make new code and old code compatible (old talking to new and new talking to old). Once only new code is running, we can safely remove sending or accepting the offending terms.

Reply:
I don't think this discussion is resolved?
> What's the real issue? Let's accept as input "master" or "primary" in one case and "slave" or "secondary" in the other, as well as send both values over the wire as output.
This will make new code and old code compatible (old talking to new and new talking to old). Once only new code is running, we can safely remove sending or accepting the offending terms.

+1, this should allow us to be backward compatible (and I'd even suggest we keep that for the life of 9 to make upgrades easy)

Reply:
That's what I added I thought @tflobbe Can you let me know where there other places where I should anticipate receiving `slave` or `master`

Reply:
> Rolling upgrade across major versions isn't supported, and I don't see any reason to bend over backwards to support that. 

Not to derail this effort at all, which I think is very important, but we *do* claim to support rolling upgrades across major versions, see for example https://lucene.apache.org/solr/guide/8_6/major-changes-in-solr-8.html#rolling-upgrades-with-solr-8. The page on upgrading a Solr cluster is really only meant for SolrCloud users (that's what was meant by "cluster" when it was written, which is actually incorrect since you can have a "cluster" without using SolrCloud), but it doesn't say not to do a rolling upgrade across major versions because we have other docs that tell you how to do exactly that.

I don't think it's a major issue though, we just need to clearly spell out the proper upgrade steps for 9.0 (which doesn't have to be part of this PR). There is going to be a lot for people to do (if they're using removed features and need to install plugins), this isn't going to add much IMO.

Reply:
> > except for the backcompat issue (which I don't think matters at all).
> 
> What's the real issue? Let's accept as input "master" or "primary" in one case and "slave" or "secondary" in the other, as well as send both values over the wire as output.
> This will make new code and old code compatible (old talking to new and new talking to old). Once only new code is running, we can safely remove sending or accepting the offending terms.

The real issue here is that the terms should've been removed years ago, or never used. They are super stressful to interact with and every time I combed through my changes here it was really frustrating. I was in college when the project was released. These terms were not used at Grinnell College. I could give a history lesson if you'd like to understand, but I think we should align with the standards put forth by the IETF and expressed in the Apache Code of Conduct at a minimum. We should also remember to prioritize community over code. We shouldn't even be looking to preserve rolling upgrades in favor of keeping such stressful nomenclature. There's so much changing in 9 already that is backward incompatible.